File: Two Pointers/11. Container With Most Water/11. Container With Most Water.js
/**
 * @param {number[]} height
 * @return {number}
 */
var maxArea = function (height) {
  let l = 0;
  let r = height.length - 1;
  let currMaxArea = 0;

  while (l < r) {
    const minHeight = Math.min(height[l], height[r]);
    const currArea = minHeight * (r - l);
    currMaxArea = Math.max(currMaxArea, currArea);

    if (height[l] < height[r]) l++;
    else r--;
  }

  return currMaxArea;
};

// Time: O(n)
// Space: O(1)

File: Two Pointers/11. Container With Most Water/11. Container With Most Water.py
class Solution:
    def maxArea(self, height: List[int]) -> int:
        l = 0
        r = len(height) - 1
        max_area = 0

        while l < r:
            min_height = min(height[l], height[r])
            curr_area = min_height * (r - l)
            max_area = max(max_area, curr_area)

            if height[l] < height[r]:
                l += 1
            else:
                r -= 1
        
        return max_area

# Time: O(n)
# Space: O(1)